### ISSUE TITLE
Ask user to automatically convert additional contacts and event participants to case [3]

### ISSUE THREAD

**Issue by kwa20** [hzi-braunschweig/SORMAS-Project#4321]
### Situation Description
Thus far, only one entity (contact or event participant) can be converted to a case at one time. Often enough however, case persons could have many contacts to source cases (same disease) or could have participated in multiple events (same disease). This is especially interesting if such contacts or event participations happened during the incubation period of the case. Taking this timeframe into consideration, all of these situations could've potentially led to the person becoming a case. Thus, all these entities of the same person should be linked.
Since users often have to handle a large workload and might forget linking all the related entities to the resulting case, a prompt asking the user if they want to link all related entities in the incubation period might heavily reduce the workload and also promote data quality.

### Feature Description
When initially converting a contact or event participant into a case, ask the user to also link associated entities (contacts, event participants) in the process. These would be entities related to the same disease and the incubation period reference would be the report date or the symptom onset date, if available and before the report date.
For example, a contact person gets converted into a case but also had four other contacts to source cases with the same disease during the incubation period of the case that the contact is being converted to. 

- When saving the case, the user would be informed that the person also had four other (confirmed) contacts to cases with that disease during the incubation period and given the choice to also convert the mentioned contacts to the same case.
- The same prompt should be displayed if an event participant is being converted into a case.
- If both, contacts and event participants exist for that person that fall into the incubation period, the prompt should mention both scenarios at the same time.
- This prompt should not only trigger upon conversion but also if a case is being created through the usual case creation process.

### Additional Information
<img width="580" alt="Screen Shot 2021-02-09 at 9 27 06 PM" src="https://user-images.githubusercontent.com/70317594/107424903-83d69080-6b1e-11eb-9512-a950bd2b32bb.png">
<img width="261" alt="Screen Shot 2021-02-09 at 9 33 11 PM" src="https://user-images.githubusercontent.com/70317594/107424922-8802ae00-6b1e-11eb-959a-db875521003f.png">

### Development Specifications
When a user is converting a contact to a case (i.e. opening a contact form and selecting "Create a case for this contact person") or creating a case for an event participant:
* Check whether there are other CONFIRMED contacts without a resultingCase of the same person and the same disease within 30 days of the contact start date (see ContactLogic.getStartDate), or other event participants without a resultingCase of the same person in events of the same disease within 30 days of the event date (start date > end date > report date, whichever is available) that the user has unpseudonymized access to
* If at least one contact or event participant is found, display a prompt to the user depending on the found entities:
    * There are X additional {disease} contacts of this person that do not have a resulting case set. Do you want to set the case you have just created as the resulting case of some or all of these contacts as well?
    * There are X additional {disease} event participants of this person that do not have a resulting case set. Do you want to set the case you have just created as the resulting case of some or all of these event participants as well?
    * There are X additional {disease} contacts and X additional {disease} event participants of this person that do not have a resulting case set. Do you want to set the case you have just created as the resulting case of some or all of these contacts and event participants as well?
* Possible options for this prompt: No, Yes, for some, Yes, for all
* If "Yes, for all" is selected, set the case as the resulting case for all of the contacts and event participants and display a tray notification notifying the user about the success of the operation ("Case set as the resulting case for all additional contacts"/"Case set as the resulting case for all additional event participants"/"Case set as the resulting case for all additional contacts and event participants")
* If "Yes, for some" is selected, open a new window that contains a hint, information about the contact/event participant that was just edited, and two lists:
    * Hint: "Please select the contacts and event participants for which you want to set the created case as the resulting case."
    * Information about the contact/event participant that was initially edited should look as it does for the respective duplicate detection
    * Grids need to be in multi-select mode
    * Contact columns can be identical to the duplicate detection columns, but without first name, last name and person ID (since those are identical anyway), also add the contact UUID (short)
    * Event participant columns: Event Participant UUID (short), Event UUID (short), Event external ID, Event title, Event status, Date of event, Event region, Event district, Event report date, involvement description (most of these columns come from the event directory, some of them have custom logic, make sure this logic is replicated)
    * The user can "Cancel" (i.e. the case is not set as the resulting case to anything) or "Confirm", in which case the case is set as the resulting case of all selected contacts and event participants
    * If the user confirmed, display a tray notification similar to the one above (but stating "for all selected additional [...]")

**Comment by kwa20** [comment]
@JaquM @bernardsilenou @MateStrysewske Any thoughts on this one?

**Comment by bernardsilenou** [comment]
@kwa20 This is a very nice one.
The only minor point i can see is performance issues since the system has to do more contrails when converting a contact to case. In addition, the system can only know plausible contact when the resulting case has been created based on the report date of the resulting case, right?
What about bulk action?, users can filter cases resulting from contacts and call method to identify plausible contact and convert them to the same case. This may not affect performance since users would do it voluntarily

**Comment by kwa20** [comment]
@bernardsilenou yes, I forgot to mention but added that the incubation period should be based on either the report date or the symptom onset date if available and before the report date.

I think your bulk action approach would also require additional filters. One reason why I would prefer a prompt would be that users would have to know about additional contacts or event participants which might not often be the case, especially for systems that many users work on. A combination of both might actually be nice since the bulk action allows retrospective "cleaning" with less effort.

**Comment by bernardsilenou** [comment]
@kwa20 
Both ways are nice to have and we would need a filter for both.
One question pls:
What about jurisdictional constrains? the user converting the contact to a case may not have full access to all the other contacts of the person. Thus blocked by system.

One option is leverage on the ticket developed by the French dev team to inform the responsible contact officer or supervisor of the other contacts to do the conversion. However, we still fall in a deadlock because the supervisor may not also have full access to the case. 
I think automatic conversion is not also good to do.
Maybe limit the feature to only contacts of that the user has full access.
I halt until I hear from u.

**Comment by MateStrysewske** [comment]
I don't know whether jurisdiction is an actual problem here. As far as I can see, this process wouldn't require any manual input from the user as it just
* Sets the case as the resulting case of all of the person's recent contacts of this disease that do not have a resulting case yet
* Sets the case as the resulting case of all of the person's event participants in events of this disease that do not have a resulting case yet

So it's basically just adding a link, and that can be done even on entities that the user doesn't have access to I think, unless this is problematic for legal reasons. In that case, we could just make sure that these processes are only performed for entities that the user has access to.

**Comment by kwa20** [comment]
I think when a user converts a contact to a case, they would also have access to the entity anyway to perform this action right? Even if the user creates a new case and selects a person that was suggested as a duplicate, the user would get access to its personal data since they are the creating user. Even if the contacts are set to another jurisdiction, linking them would be harmless since the user can access personal information through the case already.

**Comment by bernardsilenou** [comment]
@kwa20 @MateStrysewske It would be good if we limit it to only contacts that the user has full access to. Although it is rare to find a situation that the user does not have full access to the other contacts of the person. If this happen, looking from the direction of the other user, they would see that the system has convert some of their contacts to cases but they would not be able to view the case.
- Another point to consider to add in the db is to classify the conversion as user converted or system converted. Even without the validation of the user, we can still implement this issue as long as we tag the conversions as system or user converted. Do we even need to ask the user? 
- When user "unconvert" or delete the resulting case, we also need to decide how to handle this.
- It getting more complex than I thought

**Comment by kwa20** [comment]
@bernardsilenou I'd prefer the notification to raise awareness.

**Comment by bernardsilenou** [comment]
What should happen in reverse process then, notification + delete or just automatic delete and inform user of what ha happened ?

**Comment by kwa20** [comment]
@bernardsilenou Since we don't have a notification for this yet even when we only have one contact that led to the resulting case, we would have to implement this in another issue I guess. However, I think if the resulting case to those contacts was deleted (for whatever reason), the contacts should change their status back to active rather than converted to case (also cf. #1855). Thus far, the status of a contact is not changed back to 'active' when a resulting case is deleted (cf. #4353)

**Comment by kwa20** [comment]
@MartinWahnschaffe This issue is the one related to what we talked about today.

**Comment by bernardsilenou** [comment]
@kwa20 @MateStrysewske We need to finalize this ticket for the planning.
* One point I wish to know what you think is about showing the notification to inform the user rather than the user deciding if the other contacts of the person should be automatically converted to case.
*  If we allow the user to decide, then we end up with groups of persons that all contacts are converted to case and for another group, just one contact is converted.
* From @kwa20  initial explanation, this issues is to make sure that we prevent having one contact of a person converted to case and the others are still active, is this right ?

* The point to add to the initial specification is when a user convert a contact to case, we check for all recent contacts of the person + inform the user that this action would lead to the conversion of n other contacts of the person. If they proceed, we convert all, if they refuse, we convert none.
* If one of the contact is later deleted, nothing happens to others. If the resulting case is deleted, we revert back all contacts without informing user.
* What do you suggest pls?

**Comment by kwa20** [comment]
@bernardsilenou I think that is a good idea! The user will basically just be informed about the consequences of this action and merely given the choice to either proceed or cancel. We should also show a list of the associated entities that the conversion would be performed on. Note: This would be as relevant for event participants as it is for contacts.

**Comment by bernardsilenou** [comment]
@kwa20 
In this sentence 'We should also show a list of the associated entities ' do you mean like the popup with duplicate contact detection?
Contact being converted on top and others below
Maybe you can add one of those your high level mocks

**Comment by kwa20** [comment]
@bernardsilenou I thought it could look somewhat like this

<img width="599" alt="Screen Shot 2021-03-09 at 10 17 53 PM" src="https://user-images.githubusercontent.com/70317594/110539382-7a444680-8125-11eb-8f91-33bf780b1352.png">

**Comment by kwa20** [comment]
@bernardsilenou thanks! I think we need to include ep's right away independent of the follow-up that might come in the future.

**Comment by maxiheyner** [comment]
Issue tracker comment

**Comment by MateStrysewske** [comment]
@kwa20 @bernardsilenou Can you please give me a quick summary of what your proposed final solution should look like?

Also, I have serious doubts that it's a good idea to implement it like this. Wouldn't automatically setting the case as the resulting case of ALL other contacts and event participants of the person (or even forcing users to do it) potentially lead to messy and inaccurate data? Ideally, a case should only result from ONE contact and/or ONE event participant, because that would mean that we know exactly where the case got infected, right? Automatically setting it as the resulting case of all contacts and event participants would make a proper source tracing extremely tough.

@Chinedar Can you please give your input on this topic? Do you think this feature request is a good idea or do you agree with my concerns? I'm not an epidemiologist, so I'm just making a poorly educated guess here.

**Comment by kwa20** [comment]
@MateStrysewske If it would be ALL other contacts and event participants I would be concerned too, meant are actually only the ones during the incubation period of the case (dependent on the disease). If the person had multiple contacts or participated in multiple events during that time, no one could actually know where exactly the person got infected, therefore all of them would be relevant.

**Comment by JaquM** [comment]
@MateStrysewske 
A clear assignment could only be determined by a really exact sequencing, which is currently not done at Covid, therefore all actions (contacts, event participations) of the last 14 days before the onset of the disease may have led to this case. Currently, these links are lost. For this reason it is important that employees in GA cannot link all contacts and event participants individually to the same case, but they are technically supported. 
As an example, in a family of five, two people are positive, the other three are contacts to both and as soon as one of the contacts also tests positive, it is impossible to tell who was infected, as both contacts of the one now tested positive lead to the very same case.

**Comment by bernardsilenou** [comment]
@All 
All the points we have mentioned are absolutely valid. After reading the drawback of this method in @MateStrysewske s' comment https://github.com/hzi-braunschweig/SORMAS-Project/issues/4321#issuecomment-808358969 I see with Maté and think we should not implement this feature right away. We can discuss with @stephaneghozzi  also before deciding.
If you give it a second look, whether implementing this feature or not, there is no link that is lost because all contacts are indirectly linked to the resulting case through the person. 
Background infos on a possible data analysis approach:
  - In order to determine the most like contact among the many contacts of the resulting case, we need to based on a model  that would not only be base on the fact that a user convert a contact to a case but would consider:
  - Time:  @kwa20 condition of contact occurring in the incubation period of the resulting case
  - settings of these selected contacts, such as: 
        * contact type
        * frequency: single of multi-day contact
        * source case disease variant type
        * relationship with case etc
 
-  From this analysis, we can then predict the most likely contact that resulted to the case and  thus back trace the source case who may likely be the super spreader.
- We can also compare users classification and model classification to the contact that resulted to the case.
- If we then automatically convert contacts based time only,  then we bias the analysis.
- From the person view, we can present all likely contacts for the user to choose one or more and convert to case manually.
- All I can suggest for now.

**Comment by kwa20** [comment]
@MateStrysewske @bernardsilenou @JaquM I would like to elaborate on the idea behind this issue again to maybe clarify the intentions here. We should think of a consecutive approach here where we separate between the process of surveillance and analysis as two parts in Sormas.

What users do to surveil the spread of a disease next to contact tracing and follow-up is to collect data for cases and contacts in regards to the exposure of the person to the disease. Until a person becomes a case they could have had many contacts or have been taking part in many events. We can usually assume that a person had to be exposed to the disease during the incubation period but can not infer specifically which of the confirmed contacts or event participations during that time led to the disease.

When we convert contacts to cases, we don't convert them because we assessed the contact and are sure that this contact was what led to the person becoming a case but rather because this contact potentially did. We are still mainly collecting data here. Because of the possibility that all of these contacts could've resulted in this case we should convert them all to the same case.

Only then is when the analysis part comes in because now we want to see which of the contacts that potentially led to the person becoming a case had the highest potential to have been that contact. Therefore we want all these potential contacts being clearly connected to that case, especially when many users use the system at the same time and it's hard to even keep track of all the entries. If we then have all relevant contacts or participations accumulated in one place we can assess the options looking at the exposure details, contact types etc. and perform proper analyses. 

Sorry for the lengthy comment!

If the automation is the main concern, we could very well make it so users can select all or only some of the suggested entities. That shouldn't be a problem. If you agree I could provide some mock-ups for that too.

**Comment by MateStrysewske** [comment]
@bernardsilenou Yes, I'm pretty sure we should not do this during the import because it might slow it down significantly.

**Comment by bernardsilenou** [comment]
@MateStrysewske @kwa20 
- We suggest to check for individual case creation, then later add a bulk edit feature to do the same for imported cases.
- This would mean during case creation, we check for duplicate person, case, associated contracts, this may be too much of a check, tight ?

- Suggested solution is to add this feature only when converting contact or ep to case.  Then add a followup ticket with bunk action to do the same for cases created by import or not by conversion.

**Comment by kwa20** [comment]
> * allow users to decide whether or not to convert additional contacts/eps to the case, and if so
> * allow them to either automatically convert ALL of them or
> * select the contacts/eps they want to convert from two lists in a popup window

@MateStrysewske Yes, exactly! :)

> * Suggested solution is to add this feature only when converting contact or ep to case.  Then add a followup ticket with bunk action to do the same for cases created by import or not by conversion.

@bernardsilenou Yes, this is what we need for now and we should implement further options with another issue I suppose.

**Comment by JaquM** [comment]
@MateStrysewske is this all clear now for you or should we add a summary?

**Comment by MateStrysewske** [comment]
@JaquM I think it's clear, but I'll get back to you if I have any questions when I'm creating the development specifications

**Comment by MateStrysewske** [comment]
@syntakker You have used Java 11 language features in ContactFacadeEjb and EventFacadeEjb (java.util.function.Predicate.not). The target language levels of all our projects (except api and app) is Java 8, so the code doesn't compile if you've set up your projects correctly. Please get rid of the unsupported language feature and replace it with compatible code.

**Comment by syntakker** [comment]
Fixed.
